src/middlewared/middlewared/plugins/keychain.py

```diff
@@ -0,0 +1,376 @@
+                raise CallError(f"ssh-keyscan failed: {(proc.stdout + proc.stderr)!r}")
+        else:
+            raise CallError(f"ssh-keyscan failed: {(proc.stdout + proc.stderr)!r}")
+
+    @accepts(Dict(
+        "keychain_remote_ssh_semiautomatic_setup",
+        Str("name", required=True),
+        Str("url", required=True, validators=[URL()]),
+        Str("token", required=True),
+        Str("username", default="root"),
+        Int("private_key", required=True),
+        Str("cipher", enum=["STANDARD", "FAST", "DISABLED"], default="STANDARD"),
+        Int("connect_timeout", default=10),
+    ))
+    def remote_ssh_semiautomatic_setup(self, data):
+        replication_key = self.middleware.run_coroutine(
+            get_ssh_key_pair_with_private_key(self.middleware, data["private_key"]))
+
+        try:
+            client = Client(os.path.join(re.sub("^http", "ws", data["url"]), "websocket"))
+        except Exception as e:
+            raise CallError(f"Unable to connect to remote system: {e}")
+
+        with client as c:
+            if not c.call("auth.token", data["token"]):
+                raise CallError("Invalid token")
+
+            try:
+                response = c.call("replication.pair", {
+                    "hostname": "any-host",
+                    "public-key": replication_key["attributes"]["public_key"],
+                    "user": data["username"],
+                })
+            except Exception as e:
+                raise CallError(f"Semi-automatic SSH connection setup failed: {e!r}")
+
+        return self.middleware.call_sync("keychaincredential.create", {
+            "name": data["name"],
+            "type": "SSH_CREDENTIALS",
+            "attributes": {
+                "host": urllib.parse.urlparse(data["url"]).hostname,
+                "port": response["ssh_port"],
+                "username": data["username"],
+                "private_key": replication_key["id"],
+                "remote_host_key": process_ssh_keyscan_output(response["ssh_hostkey"]),
+                "cipher": data["cipher"],
+                "connect_timeout": data["connect_timeout"],
+            }
+        })
+
+    @private
+    @accepts(Dict(
+        "keychain_ssh_pair",
+        Str("remote_hostname", required=True),
+        Str("username", default="root"),
+        Str("public_key", required=True),
+    ))
+    async def ssh_pair(self, data):
+        """
+        Receives public key, storing it to accept SSH connection and return
+        pertinent SSH data of this machine.
+        """
+        service = await self.middleware.call(
+            "datastore.query", "services.services", [("srv_service", "=", "ssh")], {"get": True})
+        ssh = await self.middleware.call("datastore.query", "services.ssh", None, {"get": True})
+        try:
+            user = await self.middleware.call(
+                "datastore.query", "account.bsdusers", [("bsdusr_username", "=", data["username"])],
```

Comment:
Same here with user.query, since datastore attributes name is a implementation detail of user plugin and there is no guarantee it wont change?